### PR TITLE
[dataset] add API to update an operational dataset

### PR DESCRIPTION
**Pull request by agners**
The Dataset class has code which updates an operational dataset with information from another opertional dataset. Expose that functionality through the API. This is useful to implement the dataset REST API.

### CODE REVIEW

**Review by wgtdkp — APPROVED** [review]
LGTM 👍

File: include/openthread/dataset.h [hunk 1]
```diff
@@ -594,6 +594,15 @@ otError otNetworkNameFromString(otNetworkName *aNetworkName, const char *aNameSt
  */
 otError otDatasetParseTlvs(const otOperationalDatasetTlvs *aDatasetTlvs, otOperationalDataset *aDataset);
 
+/**
+ * This function updates an Operational Dataset with information from another Operational Dataset.
+ *
+ * @param[in]  aSrcDataset  A pointer to dataset with the new/updated information.
+ * @param[out] aDstDataset  A pointer to where the dataset will be placed.
+ *
+ */
```
Comment (abtink) [inline include/openthread/dataset.h:603]: Add comment explaining how "update" work for datasets?
```suggestion
/**
 * This function updates an Operational Dataset with information from another Operational Dataset.
 *
 * This function updates @p aDstDataset by setting all the components that are marked as present in @p aSrcDataset to
 * the same value. Other components remain unchanged in @p aDataset.
 * 
 * @param[in]  aSrcDataset  A pointer to dataset with the new/updated information.
 * @param[out] aDstDataset  A pointer to where the dataset will be placed.
 *
 */
```

**Review by abtink — APPROVED** [review]
Looks good. Thanks @agners.
Few smaller suggestions below.

**Comment by agners** [comment]
@jwhui I think this is ready to get merged :blush:

**Comment by jwhui** [comment]
@agners , thank you for submitting this PR!

One issue with `otOperationalDataset` is that it only contains attributes for Operational Dataset TLVs that are currently defined in the Thread Specification. Future Thread Specification versions may define new Operational Dataset TLVs. As a result, there would be some information loss when representing an Operational Dataset using `otOperationalDataset`. In cases where we want to update an existing Operational Dataset, it is probably safer to operate directly on `otOperationalDatasetTlvs` instead.

We should probably think about updating the OpenThread API to limit the use of `otOperationalDataset`.

Thoughts?

**Comment by agners** [comment]
My use case is the REST API, where ultimately `otDatasetSetPending` or `otDatasetSetActive` is being called. Both APIs take `otOperationalDataset`.  At that point, any additional field is anyway lost/not useful.

If we don't want to add APIs which soley operate on data structures which potentially change, then maybe supporting update variants of `otDatasetSetPending`/`otDatasetSetActive` make sense? E.g. `otDatasetUpdatePending`/`otDatasetUpdateActive`?

**Comment by jwhui** [comment]
@agners , can you provide a little more detail on your use case? In particular, are you looking to make a (partial) update to an existing Active / Pending Operational Dataset that may have been generated by some other entity? Or are you looking to explicitly set the Active / Pending Operational Dataset without regard to whatever existing Active / Pending Operational Dataset is already configured?

**Comment by agners** [comment]
My use case came from the REST API review. In my first implementation, I provided a way to update the existing dataset by checking which elements of the JSON request objects are set. This has the nice property that a new operational dataset can be created where only the properties of interest are set (e.g. network name), and the rest is initialized OTBR side with sensible defaults/random values). This also allows a similar operation like the CLI, where single commands set an element one by one.

The REST API also offers a way to write a dataset in TLV format (hex string). The first implementation just overwrote the existing dataset when a dataset in TLV format is provided.

During review, one request was to make no difference which format is used: Bascially, allow updates no matter if an operational dataset in JSON data format or in TLV  hex string format is used, see: https://github.com/openthread/ot-br-posix/pull/1682#discussion_r1071113828. Instead of parsing TLV in `response.cpp`, I'd like to reuse existing code for this feature. Hence this PR.

On a high level, currently, we are interested in two use cases:
1. Existing Thread border router provides dataset in TLV: Hook into this network by writing the active dataset in TLV format and then start the OTBR.
2. Create a new network, ideally without client side dataset TLV (hence JSON format support).

The REST API Operational Dataset API PR currently is stuck because I don't have a way to update the dataset TLV as requested by the review. If dataset TLV support proves too complicated, removing TLV entirely and parsing TLV client side might be the easier way to archive the same.

**Comment by wgtdkp** [comment]
> @agners , can you provide a little more detail on your use case? In particular, are you looking to make a (partial) update to an existing Active / Pending Operational Dataset that may have been generated by some other entity?

Exactly. The REST API wants to support updating only the Network Name with a JSON payload `{"networkname": "Thread-PAN"}`. And if we want to support such partial updates, we want to support TLV format as well.

> One issue with otOperationalDataset is that it only contains attributes for Operational Dataset TLVs that are currently defined in the Thread Specification. Future Thread Specification versions may define new Operational Dataset TLVs. As a result, there would be some information loss when representing an Operational Dataset using otOperationalDataset. In cases where we want to update an existing Operational Dataset, it is probably safer to operate directly on otOperationalDatasetTlvs instead.

What about adding API `otDatasetTlvUpdate` which accepts TLV parameters? Like this:
```
void otDatasetTlvUpdate(const otOperationalDatasetTlv *aSrcDatasetTlv, otOperationalDatasetTlv *aDstDatasetTlv);
```

**Comment by agners** [comment]
> What about adding API `otDatasetTlvUpdate` which accepts TLV parameters? Like this:
> 
> ```
> void otDatasetTlvUpdate(const otOperationalDatasetTlv *aSrcDatasetTlv, otOperationalDatasetTlv *aDstDatasetTlv);
> ```

I probably would have preferred the suggestion I've made above (`otDatasetUpdatePending`/`otDatasetUpdateActive`) as internally things will get converted anyways and these API would potentially be more useful, IMHO.

But either way works for me, as long as we are agree that this is a way forward :smile:

**Comment by agners** [comment]
Unless, `otDatasetTlvUpdate` would stay completely in TLV land, and only update based on tag?

@jwhui do you agree with this approach (adding `otDatasetTlvUpdate` as suggested by @wgtdkp?

**Comment by jwhui** [comment]
I was thinking that any APIs that directly get/set the Active/Pending Operational Datasets only accept TLV format (i.e. `otOperationalDatasetTlv`). OpenThread can provide utility functions for parsing/generating/updating TLVs using `otOperationalDataset`.

**Comment by agners** [comment]
> I was thinking that any APIs that directly get/set the Active/Pending Operational Datasets only accept TLV format (i.e. `otOperationalDatasetTlv`)


No, that is not the case, at least not right now. We have `otDatasetSetActive` (accpeting `otOperationalDataset` and `otDatasetSetActiveTlvs` (accepting `otOperationalDatasetTlv`) in https://github.com/openthread/openthread/blob/82e816b82f575283f0f3ff3c7b46ebeb87c258e0/include/openthread/dataset.h#L367-L417..

> OpenThread can provide utility functions for parsing/generating/updating TLVs using `otOperationalDataset`.

So something like:

```
void otDatasetTlvUpdate(const otOperationalDataset *aSrcDataset, otOperationalDatasetTlv *aDstDatasetTlv);
```

I could work with that, I'll just have to parse the TLV to get a `otOperationalDataset`, and use that as the update.

How would the implementation of this look like? I assume your goal is that `otOperationalDatasetTlv` does not get parsed and converted back to TLV (otherwise we end up in the same issue that fields could get lost).

**Comment by jwhui** [comment]
> No, that is not the case, at least not right now. 

Yes, I understand that the current API accepts both TLV and non-TLV. I'm proposing that we deprecate/remove the non-TLV APIs.

> How would the implementation of this look like?

I was thinking that the API would provide some helper functions to parse and manipulate an existing `otOperationalDatasetTlvs`:

- `otDatasetUpdateTlvs(otOperationalDatasetTlvs *aTlvs, const otOperationalDataset *aIn)`
- `otDatasetToTlvs(const otOperationalDataset *aIn, otOperationalDatasetTlvs *aOut)`
- `otDatasetFromTlvs(const otOperationalDatasetTlvs *aIn, otOperationalDataset *aOut)`

But in terms of updating an existing configured dataset, I would suggest the following order of calls:
1. `otDatasetGetActiveTlvs`
2. `otDatsaetUpdateTlvs`
3. `otDatasetSetActiveTlvs`

@wgtdkp @abtink , thoughts?

**Comment by abtink** [comment]
> @wgtdkp @abtink , thoughts?

Sounds good to me.
We already have `otDatasetParseTlvs()` and https://github.com/openthread/openthread/pull/8720 is adding `otDatasetConvertToTlvs()`.

**Comment by wgtdkp** [comment]
SGTM, thanks @jwhui!

@agners Will you refactor this PR to add the APIs proposed by @jwhui ?

**Comment by agners** [comment]
@wgtdkp yes I am planning to, I am not sure yet if I get around this week. I'll probably close this one and open a new one.

**Comment by wgtdkp** [comment]
> @wgtdkp yes I am planning to, I am not sure yet if I get around this week. I'll probably close this one and open a new one.

sounds good, thank you!

**Comment by agners** [comment]
With #8871 this PR is obsolete as per discussions. Thanks for implementing this @jwhui !

**Pull request closed without merging**